File: apps/web/src/types/notice.ts
export type NoticeAudienceType = "ALL" | "STUDENT" | "TEACHER" | "ADMIN";
export type NoticeType = "GENERAL" | "FEE" | "EVENT" | "HOLIDAY" | "EXAM";
export type NoticeStatus = "DRAFT" | "PUBLISHED" | "UNPUBLISHED" | "EXPIRED";

export interface NoticeTarget {
  id: number;
  tenant_id: number;
  notice_id: number;
  class_id: number | null;
  division_id: number | null;
  created_at?: string | null;
  created_by?: number | null;
}

export interface NoticeAttachment {
  id: number;
  tenant_id: number;
  notice_id: number;
  file_name: string | null;
  file_path: string | null;
  file_type: string | null;
  file_size_kb?: number | null;
  uploaded_at: string | null;
  uploaded_by?: number | null;
}

export interface Notice {
  id: number;
  tenant_id: number;
  title: string;
  description: string;
  notice_type: NoticeType;
  audience_type: NoticeAudienceType;
  status: NoticeStatus;
  publish_date: string;
  expiry_date: string | null;
  is_draft: boolean;
  is_published: boolean;
  published_at: string | null;
  unpublished_at: string | null;
  send_notification: boolean;
  created_by: number;
  created_at: string;
  updated_by: number | null;
  updated_at: string | null;
  is_deleted: boolean;
  targets: NoticeTarget[];
  attachments: NoticeAttachment[];
  /** True after the current user has opened this notice (sidebar unread). */
  is_viewed?: boolean;
}

export interface NoticeCreateTarget {
  class_id?: number;
  division_id?: number;
}

export interface NoticeCreateAttachment {
  file_name?: string;
  file_path?: string;
  file_type?: string;
  file_size_kb?: number;
}

export interface NoticeCreateRequest {
  title: string;
  description: string;
  audience_type: NoticeAudienceType;
  notice_type: NoticeType;
  publish_date?: string;
  expiry_date?: string;
  /**
   * Defaults to true on the backend when omitted.
   * CreateNotice always sends true (no admin toggle in product UI).
   */
  send_notification?: boolean;
  is_draft: boolean;
  targets: NoticeCreateTarget[];
  attachments: NoticeCreateAttachment[];
}

export type NoticeUpdateRequest = Partial<
  Pick<
    NoticeCreateRequest,
    | "title"
    | "description"
    | "audience_type"
    | "notice_type"
    | "publish_date"
    | "expiry_date"
    | "send_notification"
    | "is_draft"
    | "targets"
    | "attachments"
  >
>;

export interface NoticeListResponse {
  items: Notice[];
  total: number;
  page: number;
  size: number;
}

export interface NoticeCountResponse {
  count: number;
}

export interface NoticeStatusUpdateResponse {
  message: string;
  notice: Notice;
}

export interface NoticeDropdownOptionsResponse {
  notice_types: NoticeType[];
  audience_types: NoticeAudienceType[];
  status_types: NoticeStatus[];
}

File: apps/fastapi/app/schemas/notice.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


AudienceType = Literal["ALL", "STUDENT", "TEACHER", "ADMIN"]
NoticeType = Literal["GENERAL", "FEE", "EVENT", "HOLIDAY", "EXAM"]
NoticeStatus = Literal["DRAFT", "PUBLISHED", "UNPUBLISHED", "EXPIRED"]


class NoticeTargetItem(BaseModel):
    class_id: int | None = Field(None, ge=1)
    division_id: int | None = Field(None, ge=1)


class NoticeAttachmentItem(BaseModel):
    file_name: str | None = Field(None, max_length=255)
    file_path: str | None = Field(None, max_length=500)
    file_type: str | None = Field(None, max_length=50)
    file_size_kb: int | None = Field(None, ge=0)

    @field_validator("file_path")
    @classmethod
    def reject_embedded_content(cls, value: str | None) -> str | None:
        if value and value.strip().startswith("data:"):
            raise ValueError(
                "file_path must be a server path; upload files via the attachment endpoint"
            )
        return value


class NoticeCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: str = Field(..., min_length=1)
    audience_type: AudienceType
    notice_type: NoticeType
    publish_date: datetime | None = None
    expiry_date: datetime | None = None
    send_notification: bool = False
    status: NoticeStatus | None = None
    is_draft: bool | None = None
    targets: list[NoticeTargetItem] = Field(default_factory=list)
    attachments: list[NoticeAttachmentItem] = Field(default_factory=list)


class NoticeUpdateRequest(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = Field(None, min_length=1)
    audience_type: AudienceType | None = None
    notice_type: NoticeType | None = None
    publish_date: datetime | None = None
    expiry_date: datetime | None = None
    send_notification: bool | None = None
    status: NoticeStatus | None = None
    is_draft: bool | None = None
    targets: list[NoticeTargetItem] | None = None
    attachments: list[NoticeAttachmentItem] | None = None


class NoticeTargetResponse(BaseModel):
    id: int
    tenant_id: int
    notice_id: int
    class_id: int | None = None
    division_id: int | None = None
    created_at: datetime | None = None
    created_by: int | None = None


class NoticeAttachmentResponse(BaseModel):
    id: int
    tenant_id: int
    notice_id: int
    file_name: str | None = None
    file_path: str | None = None
    file_type: str | None = None
    file_size_kb: int | None = None
    uploaded_at: datetime | None = None
    uploaded_by: int | None = None


class NoticeResponse(BaseModel):
    id: int
    tenant_id: int
    title: str
    description: str
    notice_type: NoticeType
    audience_type: AudienceType
    status: NoticeStatus
    publish_date: datetime
    expiry_date: datetime | None = None
    is_draft: bool
    is_published: bool
    published_at: datetime | None = None
    unpublished_at: datetime | None = None
    send_notification: bool
    created_by: int
    created_at: datetime
    updated_by: int | None = None
    updated_at: datetime | None = None
    is_deleted: bool
    targets: list[NoticeTargetResponse] = Field(default_factory=list)
    attachments: list[NoticeAttachmentResponse] = Field(default_factory=list)


class NoticeListResponse(BaseModel):
    items: list[NoticeResponse]
    total: int
    page: int
    size: int


class NoticeStatusUpdateResponse(BaseModel):
    message: str
    notice: NoticeResponse


class NoticeDropdownOptionsResponse(BaseModel):
    notice_types: list[NoticeType]
    audience_types: list[AudienceType]
    status_types: list[NoticeStatus]
